src/compiler/checker.ts

```diff
@@ -8150,8 +8150,28 @@ namespace ts {
             return container === declarationContainer;
         }
 
+        function updateReferencesForTypedParameters(node: TypeNode): void {
+            const symbol = getNodeLinks(node).resolvedSymbol;
+            if (symbol) {
+                symbol.hasReference = true;
+            }
+        }
+
+        function updateReferencesForInterfaceImplementations(node: InterfaceDeclaration): void {
+            const extendedTypeNode = getClassExtendsHeritageClauseElement(node);
+            if (extendedTypeNode) {
+                const t = getTypeFromTypeNode(extendedTypeNode);
+                if (t !== unknownType && t.symbol) {
+                    t.symbol.hasReference = true;
+                }
+            }
+        }
+
         function checkIdentifier(node: Identifier): Type {
             const symbol = getResolvedSymbol(node);
+            if (symbol && !isInAmbientContext(node)) {

```

Comment:
as noted earlier we should condition these on the use of the flags. `isInAmbientContext` for example walks up the tree to reach the sourceFile. you do not need to do this for every identifier if you will never use this information.